### ISSUE TITLE
[Feature] Strided arrays with views for NestedArray?

### ISSUE THREAD

**Issue by manzt** [gzuidhof/zarr.js#15]
I see in the documentation that you've had some thoughts on strided arrays with views, rather than (or in addition to?) arrays of TypedArrays. This is something I am particularly interested in for visualization purposes.

Many visualization libraries that utilize the GPU expect a flattened array as data input, so I find myself flattening the NestedArray or trying to do things with indexing. 

Another use case would be to generate images on the fly:

```javascript
const canvas = document.getElementById("myCanvas");
const context = c.getContext("2d");

const { data, shape: [width, height] } = await z.get([0, null, null]);
const imageData = new ImageData(new Uint8ClampedArray(data), width, height);
context.putImageData(imageData, 0, 0);
```
 
Perhaps we could add this as an option:
```javascript
const slice2d = await z.get([0, null, null], { strided: true })
```

**Comment by gzuidhof** [comment]
The reason for NestedArrays here is to keep it compatible with Pyodide (which I haven't actually gotten around to combining with here), and also for implementation simpleness (although in hindsight it isn't that simple either).

I think the current NestedArray folder has become useful/complex enough to warrant separating it into it's own library I think outside of zarr. Anyway, that's a thought for another discussion. I'm happy to accept PRs for useful additions like reshaping.  For my own use cases so far this weird [[TypedArray]] hasn't actually run into important performance problems. 

---
There is one library, [**scijs ndarray**](https://github.com/scijs/ndarray) out there right now that implements strided arrays, [a lot of packages have been built on top of it](http://scijs.net/packages/). I think it's worth trying that if you want to be productive today, converting to that format should be easy. 

Scroll down on the GitHub page to the blog posts and presentations, [this one in particular](http://mikolalysenko.github.io/ndarray-presentation/#/) is worth reading through.

I have some gripes with it:
* It does not offer a numpy-like experience (well, kind of..), it leads to hard to understand code where you chain together functions like `myArray.lo(1,1).hi(3,3)`.
* It does not have typings for Typescript, and there is really no way to do it properly.
* In fact, it does not even have proper source code, if there is a bug in there you're really at a loss. The code is generated (by string concatenation) at runtime and cached, see the only [source file](https://github.com/scijs/ndarray/blob/master/ndarray.js). This scares me a bit, but it's an interesting idea though!

I think in order to be most performant one has to generate code like the author does, but that should be done at build-time of the library. It would be an interesting project to try that!

**Comment by gzuidhof** [comment]
The ideal library in my opinion would have the same interface as `NestedArray` right now, but would work by using a strided arrays under the hood (so it can be a view).

But the more I think about it.. really this should be implemented in WebAssembly, this `NestedArray`-like class should just wrap calls to WebAssembly. It would be faster, and be able to use SIMD, and I think it's possible to hide the complexity from the user. Maybe one can wrap some Rust ndarray library? Or even numpy?

Pyodide gives you this, but I think there definitely is a use-case where you don't want to ship a whole python compiler/runtime to your user (including the performance hit of using that).

**Comment by manzt** [comment]
Funny you say this! I spend yesterday playing with `wasm-bindgen` and rust [`ndarray`](https://docs.rs/ndarray/0.13.0/ndarray/). I really like this idea, however there isn't a super clean way of exposing generics directly from rust to javascript via `wasm-bindgen`. The wasm might end up being quite large to accommodate all the different bindings we need for the different typed arrays. If we can initialize a rust/wasm ndarray with a mutable ref to an array buffer, I think this would be awesome! I'm all for avoiding reinventing the wheel with javascript, especially in an age of WebAssembly and the promise of "threads" in the future. 

- 15 min tutorial on using rust + webassembly to create a JS module
https://rustwasm.github.io/book/game-of-life/introduction.html

- Useful bits of `wasm-bindgen` docs
https://rustwasm.github.io/docs/wasm-bindgen/reference/types/number-slices.html


The above would require a bit of work (that I'm interested in!), but have you given any thought to [`Tensor`](https://js.tensorflow.org/api/latest/#class:Tensor) from `@tensorflow/tfjs-core`? The main downside here is that the Tensor is immutable, but I'm not sure if that's the worst thing. Tensorflow.js supports different backends (including WASM and WebGL). 

```javascript
const { data, shape } = await zarr.get([0, null, null]);
// let data simply be a single typed array
const arr = tf.tensor(buf, shape);
arr.print()
// prints representation of immutable tensor to console
console.log(await arr.array());
// converts tensor to nested Array
```

**Comment by manzt** [comment]
https://www.tensorflow.org/js/guide/layers_for_keras_users#no_numpy_for_tensorflowjs

**Comment by gzuidhof** [comment]
Thanks for the links, I really need to take some time to learn the basics of Rust and play with this :). I have the feeling its inevitable some wasm ndarray library will pop up sooner or later, I think the wasm binary size doesn't matter that much for the kind of applications that would use it (as long as it's <50MB). 

Tensorflow's Tensor seems like it offers quite a lot of functions, it doesn't seem to support slicing in a numpy like way from what I can tell, although the alternative [isn't that much different](https://js.tensorflow.org/api/latest/#slice).  They have come a long way since keras.js!

**Comment by manzt** [comment]
Certainly! This repo is just to experiment with some of the tooling in Rust/WebAssembly and nd-arrays specifically. I am also very new to Rust. https://github.com/manzt/wasm-ndarray

I would be surprised if someone isn't working on this currently, but I can't seem to find anything.

**Comment by manzt** [comment]
With that said, rather than deciding on a library right now and wrapping that within `Zarr.js`, perhaps we could consider a way of returning a single strided TypedArray with stride metadata. This would allow the user to interface with the ndarray library they may already be using. 

const { data, shape } = await zarr.get([0, null, null]);
// let data simply be a single typed array
const arr = tf.tensor(data, shape);
// or 
const arr2 = ndarray(data, shape)

**Comment by gzuidhof** [comment]
Sure, I agree that's a good idea :). I would maybe call it `{flat: true }` or `{raw: true}`. What would the shape of a number be? `undefined`? 

Or even something like.. `{format: "nestedarray"  | "flat" | "scijs" | "tf.tensor" }`.. Although I don't see a clean way to for instance output a `tf.tensor` without embedding that into zarr's distribution and still have it be portable.

**Comment by gzuidhof** [comment]
With a different function the typings will be less complicated which is a plus, is there a big con I'm missing?

**Comment by manzt** [comment]
I like the idea of a different function, `getRaw`, and don't see a big con with this. I'm guessing most users will be expecting something like `NestedArray` when using `.get`, but using this more advanced fetch would be nice with less boilerplate (since I think it will be common with some users). 

FWIW, numpy uses `.flatten()` and `.ravel()` to return the 1D contiguous array. `flatten` returns a copy whereas `ravel` returns a view. I like `getRaw`, but perhaps we should try to conform to a similar naming convention?

**Comment by manzt** [comment]
re "shape of number":

```python
import numpy as np

arr = np.array(1)
arr.shape
# ()

arr = np.array([1])
arr.shape
# (1,)
```

Does returning an empty array make sense in javascript?

**Comment by gzuidhof** [comment]
Let's go for a separate function then :). `getFlat` vs `getRaw` I don't feel very strongly, I think the second is slightly closer to the "semantics" of what you are doing, so I'd vote for that.

For the shape of number I meant this case:
```python
>>> import numpy as np
>>> x = np.array([0,1,2,3])
>>> el = x[0]
>>> el
0
```
So in this case `z.getRaw([0])` would return a number.


There are definitely cases where you end up with a length 0 (nested)array, they are a bit weird, but that's also how numpy works.

**Comment by gzuidhof** [comment]
I'm really excited by wasm-ndarray by the way :) I'd love to contribute (not sure how much time I can promise though). This [blog post](https://hacks.mozilla.org/2019/04/pyodide-bringing-the-scientific-python-stack-to-the-browser/) has a small blurb about how they share memory between python/js in the browser and links to some related libraries (Arrow, XND).

**Comment by manzt** [comment]
Agreed. I like `getRaw` best for our purposes! 

I hope I can spend some more digging into wasm-ndarray! It was mostly an experiment to see if I could get anything working with Rust/WebAssembly and `rust-ndarray`. I've been on break for my PhD recently but will be starting again soon, so I'm not sure how much time I can afford either. However, the repo for now will serve as a place to build some ideas and could mature into something interesting I hope! I'll add you as a collaborator :)

**Comment by manzt** [comment]
Unless there is some numpy-like standard in the future, `NestedArray` + `RawArray` will cover almost all use cases. #26 closes this for now.
